docs/proposals/client-package-structure.md

```diff
@@ -0,0 +1,227 @@
+# Client: layering and package structure
+
+## Desired layers
+
+### Transport
+
+The transport layer is concerned with round-tripping requests to an apiserver
+somewhere. It consumes a Config object with options appropriate for this.
+(That's most of the current client.Config structure.)
+
+Transport delivers an object that implements http's RoundTripper interface
+and/or can be used in place of http.DefaultClient to route requests.
+
+Transport objects are safe for concurrent use, and are cached and reused by
+subsequent layers.
+
+Tentative name: "Transport".
+
+It's expected that the transport config will be general enough that third
+parties (e.g., OpenShift) will not need their own implementation, rather they
+can change the Host field to point at their own servers, etc..
+
+Action items:
+* Split out of current client package into a new package. (@krousey)
+
+### RESTClient/request.go
+
+RESTClient consumes a Transport and a Codec (and optionally a group/version),
```

Comment:
I would prefer an interface rather than a concrete type.  We've had several
places where the go http client was not usable as is (due to limitations
for proxying, http handling, and it's retry / timeout behavior).  Since a
caller could extract the transport later it's not a huge deal in the short
term.


On Nov 12, 2015, at 7:39 PM, krousey <notifications@github.com> wrote:

In docs/proposals/client-package-structure.md
<https://github.com/kubernetes/kubernetes/pull/15730#discussion_r44736922>:

> +
> +Transport objects are safe for concurrent use, and are cached and reused by
> +subsequent layers.
> +
> +Tentative name: "Transport".
> +
> +It's expected that the transport config will be general enough that third
> +parties (e.g., OpenShift) will not need their own implementation, rather they
> +can change the Host field to point at their own servers, etc..
> +
> +Action items:
> +* Split out of current client package into a new package. (@krousey)
> +
> +### RESTClient/request.go
> +
> +RESTClient consumes a Transport and a Codec (and optionally a group/version),

I think consuming an *http.Client instead of a Transport makes more sense.
The current RESTClient is constructed, then an *http.Client using a
transport is shoved in. The current RESTClient and Request are all build
around using an *http.Client and not a transport anyway. This also allows a
user to customize the client (e.g. timeouts) before building the RESTClient
instead of changing it afterwards. Seems better to me.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/15730/files#r44736922>.